clients/cli/cmd/internal/pull.go

```diff
@@ -169,33 +173,128 @@ func (target *Target) DownloadAndWriteToFile(client *phrase.APIClient, localeFil
+			fmt.Fprintln(os.Stderr, "Initiating async download...")
+		}
+		asyncDownload, _, _ := client.LocaleDownloadsApi.LocaleDownloadCreate(Auth, target.ProjectID, localeFile.ID, localeDownloadCreateParams, &localVarOptionals)
+
+		for asyncDownload.Status == "processing" {
+			if Debug {
+				fmt.Fprintln(os.Stderr, "Waiting for the files to be exported...")
+			}
+			time.Sleep(asyncWaitTime)
+			if Debug {
+				fmt.Fprintln(os.Stderr, "Checking if the download is ready...")
+			}
+			localVarOptionals := phrase.LocaleDownloadShowOpts{}
+			asyncDownload, _, _ = client.LocaleDownloadsApi.LocaleDownloadShow(Auth, target.ProjectID, localeFile.ID, asyncDownload.Id, &localVarOptionals)
+		}
+		if asyncDownload.Status == "failed" {
+			return fmt.Errorf("download failed: %s", asyncDownload.Error)
+		}
+		return fetchFile(asyncDownload.Result.Url, localeFile.Path)
+	} else {
+		file, response, err := client.LocalesApi.LocaleDownload(Auth, target.ProjectID, localeFile.ID, &localVarOptionals)
+		if err != nil {
+			if response.Rate.Remaining == 0 {
+				waitForRateLimit(response.Rate)
+				file, _, err = client.LocalesApi.LocaleDownload(Auth, target.ProjectID, localeFile.ID, &localVarOptionals)
+				if err != nil {
+					return err
+				}
+			} else {
 				return err
 			}
-		} else {
-			return err
 		}
+		return copyToDestination(file, localeFile.Path)
 	}
+}
 
+func copyToDestination(file *os.File, path string) error {
 	var data []byte
-	if file != nil {
-		data, err = ioutil.ReadAll(file)
-		if err != nil {
-			return err
-		}
-		file.Close()
-		os.Remove(file.Name())
+	data, err := io.ReadAll(file)
+	if err != nil {
+		return err
 	}
+	file.Close()
+	os.Remove(file.Name())
 
-	err = ioutil.WriteFile(localeFile.Path, data, 0644)
+	err = os.WriteFile(path, data, 0644)
 	return err
 }
 
+func fetchFile(url string, localName string) error {
+	if Debug {
+		fmt.Fprintln(os.Stderr, "Downloading file from ", url)
+	}
+	file, err := os.Create(localName)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	request, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return err
+	}
+	request.Header.Set("Authorization", "Bearer "+Config.Credentials.Token)
+	response, err := http.DefaultClient.Do(request)
+	if err != nil {
+		return err
+	}
+	defer response.Body.Close()
+	io.Copy(file, response.Body)
+	return nil
+}
+
+func asyncDownloadParams(localVarOptionals phrase.LocaleDownloadOpts) phrase.LocaleDownloadCreateParameters {
+	localeDownloadCreateParams := phrase.LocaleDownloadCreateParameters{}
+
+	localeDownloadCreateParams.Branch = localVarOptionals.Branch.Value()
+	localeDownloadCreateParams.FileFormat = localVarOptionals.FileFormat.Value()
+	localeDownloadCreateParams.Tags = localVarOptionals.Tags.Value()
+	if localVarOptionals.IncludeEmptyTranslations.IsSet() {
+		localeDownloadCreateParams.IncludeEmptyTranslations = new(bool)
+		*localeDownloadCreateParams.IncludeEmptyTranslations = localVarOptionals.IncludeEmptyTranslations.Value()
```

Comment:
Question here (mostly for my personal knowledge), since these attributes on type e.g. `IncludeEmptyTranslations` already have a datatype assigned to it in the struct defintion, is it still required to initialize them with `new(bool)`?

Reply:
I think yes, as it's defined as a bool pointer.